sparse-traits/src/operator.rs

```diff
@@ -1,41 +1,174 @@
+    fn matvec_h(&self, x: &dyn Vector, y: &mut dyn Vector) -> Result {
+        if let Some(op) = self.as_matvec_h() {
+            op.matvec_h(x, y)
+        } else {
+            Err(Error::NotImplemented)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[derive(Debug)]
+    struct SimpleVector;
+    impl Vector for SimpleVector {}
+
+    #[derive(Debug)]
+    struct SparseMatrix;
+    impl OperatorBase for SparseMatrix {
+        fn as_matvec(&self) -> Option<&dyn AsMatVec> {
+            Some(self)
+        }
+        fn as_matvec_h(&self) -> Option<&dyn AsHermitianMatVec> {
+            Some(self)
+        }
+    }
+    impl AsMatVec for SparseMatrix {
+        fn matvec(&self, _x: &dyn Vector, _y: &mut dyn Vector) -> Result {
+            println!("{self:?} matvec");
+            Ok(())
+        }
+    }
+    impl AsHermitianMatVec for SparseMatrix {
+        fn matvec_h(&self, _x: &dyn Vector, _y: &mut dyn Vector) -> Result {
+            println!("{self:?} matvec_h");
+            Ok(())
+        }
+    }
+
+    // Finite difference matrices use the following formula where f is a
+    // nonlinear function and x is a vector that we linearize around. It is not
+    // tractable to apply the transpose or Hermitian adjoint without access to
+    // the code that computes f.
+    //
+    // A y = (f(x + hy) - f(x)) / h
+    #[derive(Debug)]
+    struct FiniteDifference;
+    impl OperatorBase for FiniteDifference {
+        fn as_matvec(&self) -> Option<&dyn AsMatVec> {
+            Some(self)
+        }
+    }
+    impl AsMatVec for FiniteDifference {
+        fn matvec(&self, _x: &dyn Vector, _y: &mut dyn Vector) -> Result {
+            println!("{self:?} matvec");
+            Ok(())
+        }
+    }
+
+    /// A fallible matrix
+    #[derive(Debug)]
+    struct SketchyMatrix;
+    impl OperatorBase for SketchyMatrix {
+        fn as_matvec(&self) -> Option<&dyn AsMatVec> {
+            Some(self)
+        }
+    }
+    impl AsMatVec for SketchyMatrix {
+        fn matvec(&self, _x: &dyn Vector, _y: &mut dyn Vector) -> Result {
+            println!("{self:?} matvec");
+            Err(Error::OperationFailed)
+        }
+    }
+    #[test]
+    fn test_mult_dyn() -> Result {
+        let x = SimpleVector;
+        let mut y = SimpleVector;
+        let ops: Vec<Box<dyn OperatorBase>> =
+            vec![Box::new(SparseMatrix), Box::new(FiniteDifference)];
+        for op in ops {
+            op.matvec(&x, &mut y)?;
+            op.matvec_h(&x, &mut y).or_else(|e| {
+                eprintln!("{op:?} reported {e:?}");
+                Ok(())
+            })?;
+        }
+        Ok(())
+    }
+
+    #[test]
+    fn test_mult() -> Result {
+        let x = SimpleVector;
+        let mut y = SimpleVector;
+        let a = SparseMatrix;
+        // Static dispatch because we're using a struct that implements AsMatVec
+        a.matvec(&x, &mut y)?;
+        a.matvec_h(&x, &mut y)?;
+        Ok(())
```

Comment:
Here we can work with trait objects and concrete types using the same syntax. `test_mult_dyn` produces this output.
```
SparseMatrix matvec
SparseMatrix matvec_h
FiniteDifference matvec
FiniteDifference reported NotImplemented
```